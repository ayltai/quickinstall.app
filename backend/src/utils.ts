import { ErrorReporting, } from '@google-cloud/error-reporting';

const errors = new ErrorReporting();

export const handleError = (error : any) => {
    if (process.env.NODE_ENV === 'production') {
        errors.report(error);
    } else {
        console.error(error);
    }
};
