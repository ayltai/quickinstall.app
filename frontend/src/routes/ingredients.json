{
    "sys"   : {
        "type" : "Array"
    },
    "total" : 2,
    "skip"  : 0,
    "limit" : 100,
    "items" : [
        {
            "metadata" : {
                "tags" : [ ]
            },
            "sys"      : {
                "space"       : {
                    "sys" : {
                        "type"     : "Link",
                        "linkType" : "Space",
                        "id"       : "rljsuvr9aw29"
                    }
                },
                "id"          : "70KakHFICowgNmeSqpXTcT",
                "type"        : "Entry",
                "createdAt"   : "2024-02-22T16:56:08.307Z",
                "updatedAt"   : "2024-02-22T16:56:08.307Z",
                "environment" : {
                    "sys" : {
                        "id"       : "development",
                        "type"     : "Link",
                        "linkType" : "Environment"
                    }
                },
                "revision"    : 1,
                "contentType" : {
                    "sys" : {
                        "type"     : "Link",
                        "linkType" : "ContentType",
                        "id"       : "ingredient"
                    }
                },
                "locale"      : "en"
            },
            "fields"   : {
                "id"          : "firefox",
                "name"        : "Mozilla Firefox",
                "description" : "Firefox is a free web browser backed by Mozilla, a non-profit dedicated to internet health and privacy.",
                "category"    : "Essentials",
                "infoUrl"     : "Mozilla Firefox",
                "author"      : "Mozilla",
                "source"      : "Homebrew (Cask)"
            }
        },
        {
            "metadata" : {
                "tags" : [ ]
            },
            "sys"      : {
                "space"       : {
                    "sys" : {
                        "type"     : "Link",
                        "linkType" : "Space",
                        "id"       : "rljsuvr9aw29"
                    }
                },
                "id"          : "7ydI5TQctpoqJM8DbdyC8v",
                "type"        : "Entry",
                "createdAt"   : "2024-02-22T16:55:02.606Z",
                "updatedAt"   : "2024-02-22T16:55:02.606Z",
                "environment" : {
                    "sys" : {
                        "id"       : "development",
                        "type"     : "Link",
                        "linkType" : "Environment"
                    }
                },
                "revision"    : 1,
                "contentType" : {
                    "sys" : {
                        "type"     : "Link",
                        "linkType" : "ContentType",
                        "id"       : "ingredient"
                    }
                },
                "locale"      : "en"
            },
            "fields"   : {
                "id"          : "google-chrome",
                "name"        : "Google Chrome",
                "description" : "Chrome is the official web browser from Google, built to be fast, secure and customisable. Download now and make it yours.",
                "category"    : "Essentials",
                "infoUrl"     : "https://www.google.com/chrome/",
                "author"      : "Google",
                "source"      : "Homebrew (Cask)"
            }
        }
    ]
}
