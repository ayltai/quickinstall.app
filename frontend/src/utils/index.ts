import * as Sentry from '@sentry/react';
import { FirebaseApp, } from 'firebase/app';
import { getAnalytics, logEvent, } from 'firebase/analytics';

export const handleError = (error : any) => {
    if (import.meta.env.PROD) {
        Sentry.captureException(error);
    } else {
        console.error(error);
    }
};

export const logPageView = (app : FirebaseApp, {
    pageName,
    route,
} : {
    pageName : string,
    route    : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'page_view', {
            page_title    : pageName,
            page_location : route,
        });
    } catch (error) {
        handleError(error);
    }
};

export const logAddToCart = (app : FirebaseApp, {
    platform,
    itemId,
} : {
    platform : 'win' | 'mac',
    itemId   : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'add_to_cart', {
            items : [
                {
                    item_id : `${platform}/${itemId}`,
                },
            ],
        });
    } catch (error) {
        handleError(error);
    }
};

export const logRemoveFromCart = (app : FirebaseApp, {
    platform,
    itemId,
} : {
    platform : 'win' | 'mac',
    itemId   : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'remove_from_cart', {
            items : [
                {
                    item_id : `${platform}/${itemId}`,
                },
            ],
        });
    } catch (error) {
        handleError(error);
    }
};

export const logCheckout = (app : FirebaseApp, {
    platform,
    itemIds,
} : {
    platform : 'win' | 'mac',
    itemIds  : string[],
}) => {
    try {
        logEvent(getAnalytics(app), 'checkout', {
            items : itemIds.map((itemId : string) => ({
                item_id : `${platform}/${itemId}`,
            })),
        });
    } catch (error) {
        handleError(error);
    }
};

export const logSearch = (app : FirebaseApp, {
    platform,
    query,
} : {
    platform : 'win' | 'mac',
    query    : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'search', {
            search_term : `${platform}/${query}`,
        });
    } catch (error) {
        handleError(error);
    }
};

export const logShare = (app : FirebaseApp, {
    product,
    itemId,
} : {
    product : 'win/packages' | 'mac/packages' | 'tweaks',
    itemId  : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'share', {
            content_type : product,
            item_id      : itemId,
        });
    } catch (error) {
        handleError(error);
    }
};
