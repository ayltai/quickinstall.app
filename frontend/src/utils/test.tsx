import { createTheme, ThemeProvider, } from '@mui/material';
import { configureStore, type Store, } from '@reduxjs/toolkit';
import { render, } from '@testing-library/react';
import React, { type ReactNode, } from 'react';
import { Provider, } from 'react-redux';
import { MemoryRouter, } from 'react-router-dom';

import { appStoreApi, chocolateyApi, contentfulApi, homebrewApi, quickInstallApi, } from '../apis';
import { recipeReducer, preferenceReducer, spellReducer, } from '../states';
import { appTheme, } from '../styles';

export const createStore = (preloadedState? : unknown) => configureStore({
    preloadedState,
    reducer    : {
        recipe                          : recipeReducer,
        preference                      : preferenceReducer,
        spell                           : spellReducer,
        [ appStoreApi.reducerPath  ]    : appStoreApi.reducer,
        [ chocolateyApi.reducerPath ]   : chocolateyApi.reducer,
        [ contentfulApi.reducerPath ]   : contentfulApi.reducer,
        [ homebrewApi.reducerPath  ]    : homebrewApi.reducer,
        [ quickInstallApi.reducerPath ] : quickInstallApi.reducer,
    },
    middleware : getDefaultMiddleware => getDefaultMiddleware().concat(appStoreApi.middleware, chocolateyApi.middleware, contentfulApi.middleware, homebrewApi.middleware, quickInstallApi.middleware),
});

export const defaultStore = createStore();

export type RootState = ReturnType<typeof defaultStore.getState>;

// @ts-ignore
const theme = createTheme(appTheme);

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : Record<string, any>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => (
        <Provider store={store}>
            <MemoryRouter>
                <ThemeProvider theme={theme}>
                    {children}
                </ThemeProvider>
            </MemoryRouter>
        </Provider>
    ),
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
