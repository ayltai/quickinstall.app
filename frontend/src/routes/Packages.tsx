import type { SerializedError, } from '@reduxjs/toolkit';
import type { FetchBaseQueryError, } from '@reduxjs/toolkit/query';
import type { SearchResult, } from 'minisearch';
import React, { useCallback, useEffect, useState, } from 'react';
import { useTranslation, } from 'react-i18next';
import { useParams, } from 'react-router-dom';

import { useCaskQuery, useFormulaQuery, useIngredientsQuery, useSearchQuery, useSearchChocoQuery, } from '../apis';
import { Catalogue, CatalogueItem, } from '../components/Catalogue';
import { DetailsView, } from '../components/DetailsView';
import { HOMEBREW_REFRESH_INTERVAL, INGREDIENT_CATEGORIES, MIN_SEARCH_LENGTH, RECIPE_COMMANDS, } from '../constants';
import { useAppDispatch, useAppSelector, } from '../hooks';
import type { Ingredient, } from '../models';
import { addIngredient, removeIngredient, } from '../states';
import { handleError, } from '../utils';

export const Packages = () => {
    const { platform, } = useParams();

    const [ query,                 setQuery,                 ] = useState('');
    const [ downloadedIngredients, setDownloadedIngredients, ] = useState<Ingredient[]>([]);

    const { product, } : { product : 'win/packages' | 'mac/packages' | 'tweaks', } = useAppSelector(state => state.preference);

    const { ingredients, } : { ingredients : Ingredient[], } = useAppSelector(state => state.recipe);

    const { data : suggestedIngredients, error : ingredientsError, } = useIngredientsQuery(undefined, {
        pollingInterval : HOMEBREW_REFRESH_INTERVAL,
    });

    const { data : caskData, error : caskError, } = useCaskQuery(undefined, {
        skip            : platform === 'win',
        pollingInterval : HOMEBREW_REFRESH_INTERVAL,
    });

    const { data : formulaData, error : formulaError, } = useFormulaQuery(undefined, {
        skip            : platform === 'win',
        pollingInterval : HOMEBREW_REFRESH_INTERVAL,
    });

    const { data : searchData, error : searchError, isFetching : searchFetching, isLoading : searchLoading, } = useSearchQuery(query, {
        skip : platform === 'win' || query.length < MIN_SEARCH_LENGTH,
    });

    const { data : chocoData, error : chocoError, isFetching : chocoFetching, isLoading : chocoLoading, } = useSearchChocoQuery(query, {
        skip : platform === 'mac' || query.length < MIN_SEARCH_LENGTH,
    });

    const dispatch = useAppDispatch();

    const { t, } = useTranslation();

    const searchResultMapper = useCallback((result : SearchResult) => ({
        id          : result.id,
        name        : result.name,
        description : result.description,
        imageUrl    : result.imageUrl,
        infoUrl     : result.infoUrl,
        author      : result.author,
        source      : result.source,
    }), []);

    const handleSearch = (keywords? : string) => setQuery(keywords ?? '');

    const handleSelectionChange = useCallback((ingredient : Ingredient, selected : boolean) => dispatch(selected ? addIngredient(ingredient) : removeIngredient(ingredient)), [ dispatch, ]);

    const renderCatalogueItem = useCallback((ingredient : Ingredient) => (
        <CatalogueItem
            key={ingredient.id}
            selected={ingredients.some(item => item.id === ingredient.id)}
            item={ingredient}
            detailsView={
                <DetailsView
                    color={product === 'win/packages' ? 'info' : 'primary'}
                    description={ingredient.description}
                    infoUrl={ingredient.infoUrl}
                    source={ingredient.source} />
            }
            onSelectionChange={handleSelectionChange} />
    ), [ product, handleSelectionChange, ingredients, ]);

    const handleApiError = useCallback((error? : FetchBaseQueryError | SerializedError) => {
        if (error) {
            if ('status' in error && (error.status === 404 || error.status === 'FETCH_ERROR' || error.status === 'TIMEOUT_ERROR')) {
                console.error(error);
            } else {
                handleError(error);
            }
        }
    }, []);

    useEffect(() => {
        const newDownloadedIngredients : Ingredient[] = [];

        if (caskData) newDownloadedIngredients.push(...caskData.map(item => ({
            ...item,
            id : `${item.source}/${item.id}`,
        })));

        if (formulaData) newDownloadedIngredients.push(...formulaData.map(item => ({
            ...item,
            id : `${item.source}/${item.id}`,
        })));

        if (searchData) newDownloadedIngredients.push(...searchData.map(item => ({
            ...item,
            id : `${item.source}/${item.id}`,
        })));

        if (chocoData) newDownloadedIngredients.push(...chocoData.map(item => ({
            ...item,
            id : `${item.source}/${item.id}`,
        })));

        setDownloadedIngredients(newDownloadedIngredients);
    }, [ caskData, formulaData, searchData, chocoData, ]);

    useEffect(() => {
        handleApiError(ingredientsError);
        handleApiError(caskError);
        handleApiError(formulaError);
        handleApiError(searchError);
        handleApiError(chocoError);
    }, [ handleApiError, ingredientsError, caskError, formulaError, searchError, chocoError, ]);

    return (
        <Catalogue
            searching={searchFetching || searchLoading || chocoFetching || chocoLoading}
            suggestedItems={(suggestedIngredients ?? []).filter(ingredient => platform === 'win' ? ingredient.source === 'Chocolatey' : ingredient.source === 'Homebrew' || ingredient.source === 'Homebrew (Cask)' || ingredient.source === 'App Store')}
            downloadedItems={downloadedIngredients}
            tagLine={product === 'win/packages' ? t('packages.tagline.win') : t('packages.tagline.mac')}
            searchHint={t('packages.search.hint')}
            title={t('packages.about.title')}
            description={product === 'win/packages' ? t('packages.about.description.win') : t('packages.about.description.mac')}
            commands={RECIPE_COMMANDS}
            categories={INGREDIENT_CATEGORIES}
            itemFields={[
                'id',
                'name',
                'description',
                'imageUrl',
                'infoUrl',
                'author',
                'source',
            ]}
            loadingMessage={t('packages.loading')}
            renderCatalogueItem={renderCatalogueItem}
            searchResultMapper={searchResultMapper}
            onSearch={handleSearch} />
    );
};
