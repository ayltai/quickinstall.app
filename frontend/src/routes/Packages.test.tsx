import React from 'react';
import createFetchMock from 'vitest-fetch-mock';

import { fireEvent, render, waitFor, } from '../utils/test';
import ingredients from './ingredients.json';
import { Packages, } from './Packages';

const fetch = createFetchMock(vi);
fetch.enableMocks();

vi.spyOn(Math, 'random').mockReturnValue(0);

describe('<Packages />', () => {
    it('renders correctly', () => {
        expect(render(<Packages />)).toMatchSnapshot();
    }, 10000);

    it('searches packages', async () => {
        fetch.mockIf(request => request.url.startsWith('https://cdn.contentful.com'), request => {
            if (request.method === 'GET') return {
                status  : 200,
                body    : JSON.stringify(ingredients),
                headers : {
                    'Content-Type' : 'application/json',
                },
            };

            return {
                status : 501,
            };
        });

        const { getAllByText, getByRole, } = render(<Packages />);

        fireEvent.change(getByRole('textbox'), {
            target : {
                value : 'spotify',
            },
        });

        await waitFor(() => {
            expect(getAllByText(value => value.startsWith('spotify') || value.startsWith('Spotify')).length).toBeGreaterThan(1);
        }, {
            timeout : 10000,
        });
    }, 10000);
});
