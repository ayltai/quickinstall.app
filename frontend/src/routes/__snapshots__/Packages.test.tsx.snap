// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Packages /> > renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiStack-root css-6rg9iv-MuiStack-root"
      >
        <div
          class="MuiBox-root css-1iketvl"
        >
          <div
            class="wave"
          />
          <div
            class="wave"
          />
          <div
            class="wave"
          />
          <div
            class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4 css-1vj3kvx-MuiTypography-root"
            >
              packages.tagline.mac
            </h4>
            <div
              class="css-1izsnok"
            >
              <div
                class="css-1iz4iln"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="SearchIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
                  />
                </svg>
              </div>
              <div
                class="MuiInputBase-root MuiInputBase-colorPrimary css-8b2o5r-MuiInputBase-root"
              >
                <input
                  class="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
                  placeholder="packages.search.hint"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
        >
          <div
            class="MuiStack-root css-1xjyopk-MuiStack-root"
          >
            <div
              class="MuiBox-root css-i9gxme"
            >
              <h4
                class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-wa1b99-MuiTypography-root"
              >
                packages.about.title
              </h4>
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 css-l6aw2k-MuiTypography-root"
              >
                packages.about.description.mac
              </h6>
            </div>
            <div
              class="MuiBox-root css-hcskjo"
            >
              <div
                class="MuiBox-root css-1vraaff"
              >
                <span
                  class="index-module_type__E-SaG type"
                  style="position: absolute; width: 318px; left: 0px; top: 0px; margin-left: 28px; margin-top: 24px; color: white; font-size: 0.5rem; font-family: monospace; word-break: break-all;"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
        >
          <div
            class="MuiStack-root css-17yl5mr-MuiStack-root"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4 css-t5jrrk-MuiTypography-root"
            >
              Essentials
            </h4>
            <hr
              class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
            />
            <div
              class="MuiBox-root css-1v3caum"
            >
              <div
                class="MuiGrid-root MuiGrid-container css-xre9at-MuiGrid-root"
              />
            </div>
          </div>
          <div
            class="MuiStack-root css-17yl5mr-MuiStack-root"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4 css-t5jrrk-MuiTypography-root"
            >
              Development
            </h4>
            <hr
              class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
            />
            <div
              class="MuiBox-root css-1v3caum"
            >
              <div
                class="MuiGrid-root MuiGrid-container css-xre9at-MuiGrid-root"
              />
            </div>
          </div>
          <div
            class="MuiStack-root css-17yl5mr-MuiStack-root"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4 css-t5jrrk-MuiTypography-root"
            >
              Utilities
            </h4>
            <hr
              class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
            />
            <div
              class="MuiBox-root css-1v3caum"
            >
              <div
                class="MuiGrid-root MuiGrid-container css-xre9at-MuiGrid-root"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        aria-hidden="true"
        class="MuiBackdrop-root css-sox5kk-MuiBackdrop-root"
        style="opacity: 0; visibility: hidden;"
      >
        <div
          class="MuiStack-root css-6rg9iv-MuiStack-root"
        >
          <span
            class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-mod6k9-MuiCircularProgress-root"
            role="progressbar"
            style="width: 40px; height: 40px;"
          >
            <svg
              class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </span>
          <p
            class="MuiTypography-root MuiTypography-body1 css-1r0phtf-MuiTypography-root"
          >
            packages.loading
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiStack-root css-6rg9iv-MuiStack-root"
    >
      <div
        class="MuiBox-root css-1iketvl"
      >
        <div
          class="wave"
        />
        <div
          class="wave"
        />
        <div
          class="wave"
        />
        <div
          class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-1vj3kvx-MuiTypography-root"
          >
            packages.tagline.mac
          </h4>
          <div
            class="css-1izsnok"
          >
            <div
              class="css-1iz4iln"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
                />
              </svg>
            </div>
            <div
              class="MuiInputBase-root MuiInputBase-colorPrimary css-8b2o5r-MuiInputBase-root"
            >
              <input
                class="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
                placeholder="packages.search.hint"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      >
        <div
          class="MuiStack-root css-1xjyopk-MuiStack-root"
        >
          <div
            class="MuiBox-root css-i9gxme"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-wa1b99-MuiTypography-root"
            >
              packages.about.title
            </h4>
            <h6
              class="MuiTypography-root MuiTypography-subtitle1 css-l6aw2k-MuiTypography-root"
            >
              packages.about.description.mac
            </h6>
          </div>
          <div
            class="MuiBox-root css-hcskjo"
          >
            <div
              class="MuiBox-root css-1vraaff"
            >
              <span
                class="index-module_type__E-SaG type"
                style="position: absolute; width: 318px; left: 0px; top: 0px; margin-left: 28px; margin-top: 24px; color: white; font-size: 0.5rem; font-family: monospace; word-break: break-all;"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      >
        <div
          class="MuiStack-root css-17yl5mr-MuiStack-root"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-t5jrrk-MuiTypography-root"
          >
            Essentials
          </h4>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
          />
          <div
            class="MuiBox-root css-1v3caum"
          >
            <div
              class="MuiGrid-root MuiGrid-container css-xre9at-MuiGrid-root"
            />
          </div>
        </div>
        <div
          class="MuiStack-root css-17yl5mr-MuiStack-root"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-t5jrrk-MuiTypography-root"
          >
            Development
          </h4>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
          />
          <div
            class="MuiBox-root css-1v3caum"
          >
            <div
              class="MuiGrid-root MuiGrid-container css-xre9at-MuiGrid-root"
            />
          </div>
        </div>
        <div
          class="MuiStack-root css-17yl5mr-MuiStack-root"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-t5jrrk-MuiTypography-root"
          >
            Utilities
          </h4>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
          />
          <div
            class="MuiBox-root css-1v3caum"
          >
            <div
              class="MuiGrid-root MuiGrid-container css-xre9at-MuiGrid-root"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-sox5kk-MuiBackdrop-root"
      style="opacity: 0; visibility: hidden;"
    >
      <div
        class="MuiStack-root css-6rg9iv-MuiStack-root"
      >
        <span
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-mod6k9-MuiCircularProgress-root"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </span>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1r0phtf-MuiTypography-root"
        >
          packages.loading
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
